torch/_dynamo/variables/tensor.py

```diff
@@ -681,13 +684,49 @@ def redistribute_fn_with_prim_types(x):
                 mutable_local=variables.base.MutableLocal(),
                 **options,
             )
+
             if not self.source:
                 # Intermediary
-                unimplemented("Intermediary tensors with registered hooks - NYI")
-            else:
-                assert (
-                    fn_var.source
-                ), "Unreachable - See unimplemented for lambdas above"
+                src = fn_var.source
+                if (
+                    not src
+                    and isinstance(fn_var, variables.functions.FunctoolsPartialVariable)
+                    and fn_var.func.source
+                ):
+                    src = fn_var.func.source
+
+                if src:
+                    tx.output.guards.add(src.make_guard(GuardBuilder.ID_MATCH))
+
+                if not compiled_autograd.compiled_autograd_enabled:
```

Comment:
FWIW, this condition is kind of wrong, because whether or not compiled autograd is enabled at the time you run forwards doesn't actually say anything about whether or not compiled autograd is enabled *when you run backwards.* We may accidentally error when compiled autograd would have been enabled later (over-conservative), or vice versa (over-liberal).

To deal with the over-conservative case, a small adjustment would be to ignore the state of compiled autograd when fullgraph is requested, because it is always wrong to do unimplemented, and we ought to attempt to register hook in the graph no matter what.

To deal with the over-liberal case, it would seem to me that it would be sufficient to catch the error when running non-compiled autograd and error to the user, asking them to enable compiled autograd. I am not sure if this case is tested atm.

Reply:
I mulled this over a bunch - I think this suffices for now - we can improve this by moving it to backwards time later, but getting a graph break in forward is the most flexible option that allows for the least user facing asserts. It's the most "it just works" line of reasoning. I don't want to trace this in fwd and then raise something to the user because compiled_autograd is not yet ready for prime time - it would leave them in a situation where they just cant use compile at all.